relay-server/src/actors/project_cache.rs

```diff
@@ -468,7 +530,41 @@ impl ProjectCacheBroker {
         } = message;
 
         self.get_or_create_project(project_key)
-            .update_state(state, no_cache)
+            .update_state(state.clone(), no_cache);
+
+        // Envelopes need to remain in the queue while Relay receives invalid states from upstream.
+        if state.invalid() {
+            return;
+        }
+
+        let envelopes = self.pending_envelopes.dequeue(&project_key, |queue_key| {
+            // Pick envelopes which belong to the incoming project.
+            if project_key != queue_key.key {
+                return false;
+            }
+
+            // We return false if project is not cached or its state is invalid.
+            self.projects
+                .get(&queue_key.sampling_key)
+                // Make sure we have only cached and valid state.
+                .and_then(|p| p.valid_state())
+                .map_or(false, |s| !s.invalid())
```

Comment:
If I'm not mistaken, this logic has a flaw:

1. Imagine `self.projects` is empty.
2. A user sends a single envelope with a project key and a distinct sampling key.
3. The project state for the event's "own" project comes in first.
4. We leave the envelope on the queue, because the sampling state is not in `self.projects` yet.
5. The project state for the sampling key comes in, but we leave the envelope on the queue because `project_key != queue_key.key`.
6. No new envelopes come in, and the envelope stays on the queue forever.

I think this problem can be overcome by treating the two partial keys symmetrically here. That is:

```suggestion
            let other_key = if queue_key.key == project_key { queue_key.sampling_key } else { queue_key.key };

            // We know that `project_key`s state has been loaded and is valid, so now check the other one:
            self.projects
                .get(&other_key)
                // Make sure we have only cached and valid state.
                .and_then(|p| p.valid_state())
                .map_or(false, |s| !s.invalid())
```

This way, we will always deque if and only if both states are loaded.